### ISSUE TITLE
[redux] Support redux-thunk

### ISSUE THREAD

**Issue by remko** [flowtype/flow-typed#574]
The type definition of `Dispatch` states:

       declare type Dispatch<A: { type: $Subtype<string> }> = (action: A) => A;

Am I right that this doesn't cover support for `redux-thunk` middleware, where functions are dispatched instead of objects? Could this be added?

**Comment by mtford90** [comment]
I changed the definition to the following in my `flow-typed/npm/redux_v3.x.x.js`:

```
declare type Dispatch<A: { type: $Subtype<string> }> = (action: A | (Dispatch<A> => mixed)) => A;
```

**Comment by aaronjensen** [comment]
redux-thunk is not part of redux, it should not be supported as part of the redux libdefs.

Instead, you can do something like this: https://github.com/agentcooper/react-redux-flow-example/blob/master/src/types/index.js#L20

**Comment by satya164** [comment]
That doesn't work. We're exporting a different `Dispatch`, but the dispatch definition inside redux's dispatch is not changed

**Comment by satya164** [comment]
This doesn't work. There are still Flow errors if I use `bindActionCreators`

**Comment by rafaelkallis** [comment]
Is anybody working on this?

**Comment by satya164** [comment]
@rafaelkallis This is the best I could do https://blog.callstack.io/type-checking-react-and-redux-thunk-with-flow-part-2-206ce5f6e705#.jseb9mo7r

**Comment by rafaelkallis** [comment]
Hey @satya164 I've read your blog post earlier while browsing the webs, great work. You should make a PR; the community would appreciate it!

**Comment by satya164** [comment]
@rafaelkallis yeah, just need to figure out how to make `bindActionCreators` work with flow

**Comment by rafaelkallis** [comment]
@satya164 is [this](https://github.com/flowtype/flow-typed/blob/master/definitions/npm/redux_v3.x.x/flow_v0.33.x-/redux_v3.x.x.js#L43-L47) what you're looking for?

**Comment by satya164** [comment]
it's tricky because we are not declaring types for `bindActionCreators` like we do for store, and it doesn't infer the correct thing :(

**Comment by hmeerlo** [comment]
Hi, I got it working, please see my sample repo: https://github.com/hmeerlo/redux-thunk-flow

**Comment by satya164** [comment]
I think better to have a way to specify a types for `bindActionCreators` rather than needed to modify the types for redux. It'll not likely get accepted here because thunk is not a part of redux, so the types will be wrong for people who don't use thunk.

You can never have 100% flow coverage anyways, reducers are terrible at type checking.

**Comment by hmeerlo** [comment]
I agree that it is not the best solution, but for me it is the only working way to use flow with redux-thunk. I have tried the approach with a new definition for bindActionCreators but that opened Pandora's box.

**Comment by ktosiek** [comment]
Would having an optional 3rd argument to Store with the type of "dispatchable" action help? This way it would be possible to separate what is accepted by middleware from what's accepted by the reducer. It would be even better for Middleware to take 2 separate actions, but I don't know how that'd work with applyMiddleware.

**Comment by mpolichette** [comment]
I kind of agree with @ktosiek's idea, but basically the _dispatchable_ thing is really just your own type definitions for `Action`...  We already get to pass `Action` into the `Store` so why cant we just use that.  

The main issue is that the definition for `Dispatch` adds the extra requirement on the `Action` we pass in, specifically, it adds the requirement that it has a property, `type`, which is a string.   I think, this  requirement is technically only a requirement of the "actions" passed to the reducers, not the actions passed to `dispatch`.  I know this makes two separate `Action` types, but I think the blame there is the fault of `middleware` :/

Normally I'm all for strictness, but in this case I would make the argument that since `middleware` is a full feature of redux, and since middleware is allowed to intercept "actions" before they get to the reducers, I would argue that dispatch shouldn't have the extra requirement. 

My vote would be to remove the $Subtype declaration: [<A: { type: $Subtype<string> }>](https://github.com/flowtype/flow-typed/blob/master/definitions/npm/redux_v3.x.x/flow_v0.33.x-/redux_v3.x.x.js#L10).

**Comment by pronebird** [comment]
Just make it possible to pass your own `Dispatch` function.

```
declare type Dispatch<A: { type: $Subtype<string> }> = (action: A) => A;
declare type Store<S, A, D = Dispatch<A>> = {
  dispatch: D;
}
```

**Comment by mikeplus64** [comment]
Removing the `{type: $Subtype<string>}` is a bit dishonest though. It *is* required by `redux` (though redux doesn't actually use it anywhere). See https://github.com/reactjs/redux/blob/e58e207213161042cb5853ddb6acb0d65034c3b0/src/createStore.js#L157

That said in a few `react-redux` projects I've got I just use this type for `dispatch`, which works pretty nicely:

```javascript

type State = {||};

type Thunk<R> = ((dispatch: Dispatch, getState: () => State) => R);
type Action = A | B | ...;

type Dispatch
  = (<R> (action: Thunk<R>) => R)
  & (<A: Action & { type: $Subtype<string> }> (action: A) => A);
```

**Comment by pronebird** [comment]
@mikeplus64 `Store.dispatch` is defined as `dispatch: Dispatch<A>`. There is no way to remove the requirement for `type` key.

The best imo if custom `Dispatch` could be provided along with generics with fallback to plain `Dispatch<A>`.

**Comment by pronebird** [comment]
I put together a patch that adds `Dispatch` parameter to override the default `Dispatch`.

**flow-typed/npm/redux_v3.x.x.js**: https://gist.github.com/pronebird/68a43881ff5774c6095ce14a9e51460f
**diff**:
https://gist.github.com/pronebird/3d735e02691a4d5b67ac017f81d1799f

The store can be setup as following:

```
type Action = { type: "foo" } | { type: "bar" };
type Thunk = (dispatch: Dispatch, getState: GetState) => any;

type State = { .. };
type GetState = () => State;

type Dispatch = (action: Action | Thunk) => any;
type MyStore = Store<State, Action, Dispatch>

const store: MyStore = createStore(rootReducer, enhancer);

// both should not trigger any flow errors
store.dispatch({ type: 'foo' });
store.dispatch((d, getState) => {});
```

In my source `D` is obligatory however default value with default `Dispatch` can be used for use-cases when default `Dispatch`.

I would send a PR but I need somebody to approve that.

**Comment by mrm007** [comment]
If anyone is looking for a way to type check their bound action creators, this is what I did:

```js
// @flow
type State = {| /* whatever */ |};

type Action = { type: $Subtype<string> };
type _ThunkAction<R> = (dispatch: Dispatch, getState?: GetState) => R;
type ThunkAction = _ThunkAction<any>;
export type Dispatch = (action: Action | ThunkAction) => any;
export type GetState = () => State;
type ThunkActionCreator<Args, R, F: (...args: Args) => _ThunkAction<R>> = (...args: Args) => R;
export type BoundAction<F> = ThunkActionCreator<*, *, F>;

function getSomething(value: number, other: string) {
  return (dispatch: Dispatch) => {
    dispatch({ type: 'get', value, other });
    return 1;
  };
}

// @connect(null, { getSomething })
class App extends React$Component {
  props: {
    getSomething: BoundAction<typeof getSomething>,
  };

  render() {
    // type checking works for arguments AND return value
    const result: string = this.props.getSomething('1', '2');
    return <div>{result}</div>;
  }
}

/* OUTPUT:
27:     const result: string = this.props.getSomething('1', '2');
                               ^ number. This type is incompatible with
27:     const result: string = this.props.getSomething('1', '2');
                      ^ string
27:     const result: string = this.props.getSomething('1', '2');
                                                       ^ string. This type is incompatible with the expected param type of
12: function getSomething(value: number, other: string) {
                                 ^ number
*/
```

[Try it out on flow.org/try](https://flow.org/try/#0PTAEAEDMBsHsHcBQAXAngBwKagMrIIbLYC8oA3gD6jABUo8AFoZgG6YBOoNYFAvgNyIUGbAEEAxsgCWsAHahSZUGiwAuUABIcAVwBGKzAB4AzsnZTZAcwB8oAcKygA+gBUG22QGsJ0uYYBKtqQAFAAmUsbohOIM6gAiEVHIMQA0oJaYyHjMAPzqAOKZ2UQAlAq2-oIGoG4e3pIy8qSu7l4+jYb4sqjWgpgAHuiw7MjKIqAJkdEMCqDB+A1y6u1yoFS1bYuyZcS2Xah9g8Oj1YVZBESzwTu2xZhV4xv1vrIAwuyYhMOGouyWxml-GkAGLqYIAOkh+D+xmWMJuzieK1kAWsQTmkPB0P+cP+CMqiAGQxGY0cACFYB5QsjDMD0Uitu9PshvjQ0mzQHTBIhIB4tukirAALaZBgWSzBFj4aDaTDqWTaIW6DhpWDIBgcdSmcxWMpkRCgUAfZDadjyMKJabxS3JBgI-WGw3hKa24JKAzqADkGWQnrSUplmFV6o4dhKgkdRsypvkAEYI3ZBLwhCAIOI5LJMJJggroNA0kofThhaLxWHEOJoPhjMZQKJ0OhQAMiLJQrX-J9JBpXsKhpnZKMHaB0OxYOhYeQDY6iyX1eL1BSqTSDLBIAKsrOxVZrCkp-Ypx9WxxrpPI6nqjEs54y-Bhp5a5BhqBsYrMAPa6IAHJxKMms2gANZSnQ101kUwo2MbRoGQLUzDLUg52McERzHJCZxFOcrGCT1Yz9UBPQAJk9cNgN-GNQEMcIWGsMgPkg6DeEMYAqN6PdEGTRBaFAAB5ABVFwAAV+NURACIAdnUSNQPAuioJg0BtXg5QxSQlDx3BdDSywnC8MI4iE0jQyjOM4yAD1QAVJUOHBGoVNJbAIlACx0yFJIpF0aBsHgKR1VEiTDOk0ZZOg2CdUsWZEOQ0d1M0zCJR0tI9JIkyUsM8zFKsPzJMdQKILk0KlMitS0MFDCt3i3DEqI5LUtqur6pM9K4KsGy3Ec6pHOc3tCHczz6B8mYQybQYsyIUJh2hfAhXs0BV0QWMCPUXlZH5WLyslaVZXlRVlXYYMNXYArdVPBrTsdczLN2xBuCAA)

**Comment by Hypnosphi** [comment]
Looks like redux `Middleware` isn't typed correctly. Currently, it looks like this:

```
declare export type Middleware<S, A, D = Dispatch<A>> =
    (api: MiddlewareAPI<S, A, D>) =>
      (next: D) => D;
```

The problem here is `(next: D) => D` part which assumes that middleware can't actually change `dispatch` signature. This is not true for quite a lot of middlewares including thunk.

Actual signature for thunk, basing on [source](https://github.com/gaearon/redux-thunk/blob/master/src/index.js) is roughly that:
```
// NA = next action, NR = next return value, R = return value

(api: MiddlewareAPI<S, A, D>) =>
      (next: NA => NR) => <R>((D => R) | NA) => R | NR;
```

Ideally, it should be not `<R>((D => R) | NA) => R | NR` but rather `(<R>(D => R) => R) & (NA => NR)` (return R for thunks, apply next for others), but flow doesn't seem to understand such signatures:
https://flow.org/try/#0C4TwDgpgBAYlC8UAUBnYAnAlgOwOYID4psBXAWwCMJ0BKKAMmVMusKjSzxoChuBjAPbY0UAGYAuWAigAPNqEgDRshPEQByZlXTqoAfigBlDDlxIZdSQDly28zSA

**Comment by pronebird** [comment]
Redux Middleware is types correctly. You only need to type your Store with a custom Dispatch parameter which can be union.

**Comment by ktosiek** [comment]
It's correct, but could be more precise - for example, the types could enforce the order of thunk and another middleware that depends on thunk.

**Comment by Hypnosphi** [comment]
> a middleware that depends on thunk cannot be used without thunk somewhere inside `next`

It's worse: the thunk itself cannot be used without thunk somewhere inside `next`, because of `D => D` part

**Comment by remhume** [comment]
I'm currently using this type for Dispatch.  
It's not accurate at all and frankly a hack, but at least it works so far. I'm only leaving this here for people who also need a quick fix.

I'm personally not concerned for the missing `{ type: $Subtype<string> }` check, because i have a function that creates my action creators.

The proposed types above (using Intersections) would always throw errors for me in weird ways.

```jsx
type Dispatch = <Ret, ThunkAction: (dispatch: Function, getState: Function) => Ret>(a: ThunkAction | Ret) => Ret
```

**Comment by AndrewSouthpaw** [comment]
> I'm personally not concerned for the missing { type: $Subtype<string> } check, because i have a function that creates all my action creators which checks this.

Ooh! Can you share? I'm curious to see.
